runbot/container.py

```diff
@@ -174,15 +187,34 @@ def _docker_run(cmd=False, log_path=False, build_dir=False, container_name=False
     rc_path = os.path.join(build_dir, '.odoorc')
     with open(rc_path, 'w') as rc_file:
         rc_file.write(rc_content)
-    docker_command.extend(['--volume=%s:/home/odoo/.odoorc:ro' % rc_path])
+    volumes.update({rc_path: {'bind': '/home/odoo/.odoorc', 'mode': 'ro'}})
 
+    ports = {}
     if exposed_ports:
         for dp, hp in enumerate(exposed_ports, start=8069):
-            docker_command.extend(['-p', '127.0.0.1:%s:%s' % (hp, dp)])
+            ports.update({f'{dp}/tcp': ('127.0.0.1', hp)})
+
+    ulimits = [docker.types.Ulimit(name='core', soft=0, hard=0)]
     if cpu_limit:
-        docker_command.extend(['--ulimit', 'cpu=%s' % int(cpu_limit)])
-    docker_command.extend([image_tag, '/bin/bash', '-c', "%s" % run_cmd])
-    subprocess.Popen(docker_command, stdout=logs, stderr=logs, preexec_fn=preexec_fn, close_fds=False, cwd=build_dir)
+        ulimits.append(docker.types.Ulimit(name='cpu', soft=cpu_limit, hard=cpu_limit))
+
+    logs.close()
+    docker_client = docker.from_env()
+    docker_client.containers.run(
+        image_tag,
+        name=container_name,
+        volumes=volumes,
+        shm_size='128m',
+        mem_limit=memory,
+        ports=ports,
+        ulimits=ulimits,
+        environment=env_variables,
+        init=True,
+        command=['/bin/bash', '-c',
+                 f'exec &>> /data/buildlogs.txt ;{run_cmd}'],
+        auto_remove=True,
+        detach=True
+    )
```

Comment:
One part of the  "file based docker state" is that in some cases, even if the container was started, the returned status was that it wasn't (asking to fast after the docker run)

Hi would propose to test that with the docker api to check if it may be possible to simplify that.


```suggestion
    container = docker_client.containers.run(
        image_tag,
        name=container_name,
        volumes=volumes,
        shm_size='128m',
        mem_limit=memory,
        ports=ports,
        ulimits=ulimits,
        environment=env_variables,
        init=True,
        command=['/bin/bash', '-c',
                 f'exec &>> /data/buildlogs.txt ;{run_cmd}'],
        auto_remove=True,
        detach=True
    )
    if container.status != 'running':
            _logger.error('Container started but status is not running:  %s' % container.status)
```